api/lock.go

```diff
@@ -165,25 +154,36 @@ WAIT:
 	if pair != nil && pair.Flags != LockFlagValue {
 		return nil, ErrLockConflict
 	}
-	if pair != nil && pair.Session != "" {
+	if pair != nil && pair.Session != "" && pair.Session != l.lockSession {
 		qOpts.WaitIndex = meta.LastIndex
 		goto WAIT
 	}
 
-	// Try to acquire the lock
-	lockEnt := l.lockEntry(l.lockSession)
-	locked, _, err := kv.Acquire(lockEnt, nil)
-	if err != nil {
-		return nil, fmt.Errorf("failed to acquire lock: %v", err)
-	}
+	if pair == nil || pair.Session == "" {
+		if l.lockSession != "" {
```

Comment:
Wait, I think I see what you're talking about now. That logic doesn't make sense because we always have a l.lockSession != nil. Hmm. I remember there was an invalid session issue. Let me look further.